Filter snv and sv variants by chromosome coordinates

northwestwitch opened the pull request:
Fix #1303.

I meant to add form validators so that chromosome, start and stop are checked before submitting the search. But after spending so many hours on this I've realized that we should first fix variants and sv_variants view because the code there is very long and convoluted. After fixing the methods it would be so much easier adding the form validation

**How to test**:
1. Open snv variants page and perform search using chromosome, start and stop parameters in search filter
1. Open sv variants page and perform search using chromosome, start and stop parameters in search filter

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by dNil
- [x] tests executed by Travis, and to a humble degree by dNil

dnil commented:
Oh, and dont forget to add an sv coords function to the adapter - query! The SNV one will not work... It would only find SVs *fully contained* in the search area; what you want is with one of the breakpoints in the given window.

northwestwitch commented:
> The SNV one will not work.

You mean the SV one right?

dnil commented:
> > The SNV one will not work.
> 
> You mean the SV one right?

:+1: The existing SNV interval query function will not do the trick for an interval query with general SVs.

northwestwitch commented:
Ready for review, whoever has time, with a preference for @dnil because it contains changed code in the monster variant query function

dnil commented on scout/adapter/mongo/query.py:
```diff
@@ -324,17 +338,56 @@ def coordinate_filter(self, query, mongo_query):
+        # Variant             xxxxxxxxxxxxxx
+        coordinate_query = {
+            "$or": [
+                { "end": { "$gte": int(query['start']), "$lte": int(query['end']) }}, #1
```
Hm, you still need to check chromosome as well. BNDs appear with different start and end chromosomes.

northwestwitch replied:
Right! On it!

dnil commented on scout/adapter/mongo/query.py:
```diff
@@ -163,7 +164,15 @@ def build_query(self, case_id, query=None, variant_ids=None, category='snv'):
+                    if category == 'sv':
+                        coordinate_query = [self.sv_coordinate_filter(query, mongo_query)]
+                    else:
+                        self.coordinate_filter(query, mongo_query)
```
Its a little strange that the coordinate_filter and sv_coordinate_filter have different calling mechanics. Either unify or change the name of the sv_coordinate_filter to sth like sv_coordinate_query.

northwestwitch replied:
I'd rather not unify them because they work in different ways, I'll rename the sv one to sv_coordinate_query

dnil replied:
✅

dnil commented on scout/adapter/mongo/query.py:
```diff
@@ -324,17 +338,56 @@ def coordinate_filter(self, query, mongo_query):
+                {   "$and": [
+                    {"position": {"$lte": int(query['start'])} },
+                    {"end": {"$gte": int(query['end'])} }
+                ]}  #4
```
Otw good start! :+1: for the numbered pictograms!

dnil commented on tests/adapter/mongo/test_query.py:
```diff
@@ -539,146 +578,151 @@ def test_build_range(adapter):
     assert mongo_query['position'] == {'$lte': end}
     assert mongo_query['end'] == {'$gte': start}
 
-def test_get_overlapping_variant(populated_database, parsed_case):
```
Eh, sorry, where did the original overlapping variant test go?

northwestwitch replied:
I replaced the code of it with new code, because the test was very slow

northwestwitch commented:
Fixed according to comments!

northwestwitch commented:
Now it's fixed!

dnil commented:
Way to go! 

There seem to be a couple of things to think about still, esp with regards to form validation. Here's searching with a backwards variant (end < start):

<img width="1633" alt="Screenshot 2019-10-24 at 17 42 42" src="https://user-images.githubusercontent.com/758570/67502308-f8fe2800-f685-11e9-88cf-e463472cb55f.png">

Not exactly the behaviour I might expect?

Also leaving the chromosome empty while entering coordinates should perhaps not be allowed? Currently that behaviour is sort of undefined?

<img width="1188" alt="Screenshot 2019-10-24 at 17 47 11" src="https://user-images.githubusercontent.com/758570/67507256-d8869b80-f68e-11e9-9e11-f19d98cd3234.png">

Then just semantics, but I personally prefer *end* over *stop* for the labels: the latter is so tightly conjoined with aa end for me..

I'm not so bothered by it, but I have a feeling you would prefer this to look different, right?
<img width="303" alt="Screenshot 2019-10-24 at 17 38 42" src="https://user-images.githubusercontent.com/758570/67501834-3ada9e80-f685-11e9-8825-c47632db530e.png">

dnil commented on scout/server/blueprints/variants/forms.py:
```diff
@@ -70,7 +70,7 @@ class FiltersForm(FlaskForm):
     chrom = TextField('Chromosome', [validators.Optional()])
     start = IntegerField('Start position', [validators.Optional(), IntegerField])
-    end = IntegerField('Stop position', [validators.Optional(), IntegerField])
+    end = IntegerField('End position', [validators.Optional(), IntegerField])
```
:+1:

northwestwitch replied:
This is just in place for when we introduce Flask_WTF validate_on_submit. The code is not validated yet.

dnil replied:
But it is "End" position! Validation can wait. 😄 Besides, there is temporary validation in place - which I ofcourse though you were already calling here as well.. X-)

northwestwitch replied:
> But it is "End" position! Validation can wait. 😄 Besides, there is temporary validation in place - which I ofcourse though you were already calling here as well.. X-)

I changed into End, didn't you like it more than Stop?

dnil replied:
Yes! Hence thumbs up! 😆

northwestwitch replied:
Ah ok, now I see it!

dnil commented on scout/server/blueprints/variants/templates/variants/sv-variants.html:
```diff
@@ -193,6 +193,35 @@
+          alert("Both start and end coordinates are required");
+          return false;
+        }
+        else if( (isNaN(start) || isNaN(end)) || end<start ){
```
:+1:

dnil commented on scout/adapter/mongo/query.py:
```diff
@@ -356,36 +355,45 @@ def sv_coordinate_filter(self, query, mongo_query):
+        coordinate_query = None
+        chromosome_query = { '$or' : [
+            {'chromosome' : query['chrom'] },
+            {'end_chrom' : query['chrom']}
```
:+1:

northwestwitch commented:
> Way to go!
> 
> There seem to be a couple of things to think about still, esp with regards to form validation. Here's searching with a backwards variant (end < start):
> 
> <img alt="Screenshot 2019-10-24 at 17 42 42" width="1633" src="https://user-images.githubusercontent.com/758570/67502308-f8fe2800-f685-11e9-88cf-e463472cb55f.png">
> 
> Not exactly the behaviour I might expect?

This happens because the form validation is not yet in place. I was waiting to re-write the variants view to organize it better before introducing form filtering, otherwise the code would be even more messy. But you are right, you don't want users to experience this type of errors, so I've introduced a temporary javascript check (only for chr, start and end form fields) on form submission. We'll remove it when we'll validate via WTF.
This type of error should be gone now

> Also leaving the chromosome empty while entering coordinates should perhaps not be allowed? Currently that behaviour is sort of undefined?
> 
> <img alt="Screenshot 2019-10-24 at 17 47 11" width="1188" src="https://user-images.githubusercontent.com/758570/67507256-d8869b80-f68e-11e9-9e11-f19d98cd3234.png">

Fixed this using the same javascript validation an above.

> Then just semantics, but I personally prefer _end_ over _stop_ for the labels: the latter is so tightly conjoined with aa end for me..

End is more compliant with the rest of the code as well, so I've renamed Stop to End.

> I'm not so bothered by it, but I have a feeling you would prefer this to look different, right?
> <img alt="Screenshot 2019-10-24 at 17 38 42" width="303" src="https://user-images.githubusercontent.com/758570/67501834-3ada9e80-f685-11e9-8825-c47632db530e.png">

Fixed, I've forgotten colspan field in the td!

dnil commented on scout/server/blueprints/variants/templates/variants/variants.html:
```diff
@@ -93,7 +92,7 @@
             {% else %}
               <tr>
-                <td colspan="">
+                <td colspan="10">
```
😄

dnil commented on CHANGELOG.md:
```diff
@@ -21,6 +21,7 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 - Variants can be marked as partial causative if phenotype is available for case.
 - Show a frequency tooltip hover for SV-variants.
 - Added support for LDAP login system
+- Search snv and variants by chromosomal coordinates
```
```suggestion
- Search snv and structural variants by chromosomal coordinates
```

northwestwitch commented:
Shall I merge?

dnil commented:
Yes lets merge -  its tested and approved. It will see even more integration testing before we bump version and deploy.
